components/raftstore/src/store/snap.rs

```diff
@@ -2009,46 +2008,53 @@ impl SnapRecvConcurrencyLimiter {
         SnapRecvConcurrencyLimiter {
             limit: Arc::new(AtomicUsize::new(limit)),
             ttl_secs,
-            timestamps: Arc::new(Mutex::new(VecDeque::with_capacity(limit))),
+            timestamps: Arc::new(Mutex::new(HashMap::with_capacity_and_hasher(
+                limit,
+                Default::default(),
+            ))),
         }
     }
 
     // Attempts to add a snapshot receive operation if below the concurrency
     // limit. Returns true if the operation is allowed, false otherwise.
-    pub fn try_recv(&self) -> bool {
+    pub fn try_recv(&self, region_id: u64) -> bool {
         let mut timestamps = self.timestamps.lock().unwrap();
         let current_time = Instant::now();
         self.evict_expired_timestamps(&mut timestamps, current_time);
 
         let limit = self.limit.load(Ordering::Relaxed);
         if limit == 0 {
-            // 0 means no limit. In that case, we avoid pushing into the
-            // VecDeque to prevent it from growing indefinitely.
+            // 0 means no limit. In that case, we avoid inserting into the hash
+            // map to prevent it from growing indefinitely.
             return true;
         }
 
-        if timestamps.len() < limit {
-            timestamps.push_back(current_time);
+        // Insert into the map if its size is within limit. If the region id is
+        // already present in the map, update its timestamp.
+        if timestamps.len() < limit || timestamps.contains_key(&region_id) {
+            timestamps.insert(region_id, current_time);
             return true;
         }
         false
     }
 
-    fn evict_expired_timestamps(&self, timestamps: &mut VecDeque<Instant>, current_time: Instant) {
-        while let Some(&timestamp) = timestamps.front()
-            && current_time.duration_since(timestamp) > Duration::from_secs(self.ttl_secs)
-        {
-            timestamps.pop_front();
-        }
+    fn evict_expired_timestamps(
+        &self,
+        timestamps: &mut HashMap<u64, Instant>,
+        current_time: Instant,
+    ) {
+        timestamps.retain(|_, timestamp| {
+            current_time.duration_since(timestamp.clone()) <= Duration::from_secs(self.ttl_secs)
```

Comment:
Maybe we should add a log for this unusual case?

Reply:
added a warning for it